shared/subspace-data-retrieval/src/object_fetcher/partial_object.rs

```diff
@@ -0,0 +1,736 @@
+    /// - they aren't at the end of a segment
+    /// - they are at the end of a segment, but all potential padding bytes are non-zero
+    /// - we've checked the non-zero ignored padding lengths, and they were all invalid
+    ///
+    /// Panics if the object has no valid lengths left.
+    fn has_padding(&self) -> bool {
+        !self.suffix_data.is_empty()
+            || self.lengths.len() > 1
+            || self.largest_ignored_padding_length() > 0
+    }
+
+    /// Returns the longest padding length the object could possibly have.
+    /// Adding padding can change the object length, but this is always returns the longest
+    /// possible amount of padding bytes.
+    ///
+    /// Panics if the object has no valid lengths left.
+    fn largest_ignored_padding_length(&self) -> usize {
+        self.lengths
+            .iter()
+            .map(|length| length.ignored_padding_length)
+            .max()
+            .expect("other methods return an error if lengths becomes empty; qed")
+    }
+
+    /// Check the hash against all possible objects with enough data.
+    /// If a possible object's data doesn't match the mapping hash, its length is removed.
+    ///
+    /// Returns valid object data, `Ok(None)` if we need more data, or an error if all possible
+    /// objects have invalid hashes.
+    ///
+    /// Panics if it is called after there are no possible lengths left.
+    /// (The previous call returned an error, so the caller should have stopped checking.)
+    pub fn check_available_objects(
+        &mut self,
+        mapping: GlobalObject,
+    ) -> Result<Option<Vec<u8>>, Error> {
+        trace!(?mapping, ?self, "checking available objects");
+
+        // Try to decode the shortest object(s), until we don't have enough data.
+        loop {
+            let outcome = Vec::<u8>::decode(&mut self.shortest_object_data());
+
+            trace!(
+                checked_length = ?self.lengths.first(),
+                data = ?hex::encode(self.shortest_object_data().0.collect::<Vec<u8>>()),
+                outcome = ?outcome.as_ref().map(hex::encode),
+                "checking object with length",
+            );
+
+            let Ok(data) = outcome else {
+                // Tell the caller we need more data, because the remaining lengths are longer.
+                return Ok(None);
+            };
+
+            let data_hash = blake3_hash(&data);
+
+            if data_hash == mapping.hash {
+                return Ok(Some(data));
+            } else {
+                // If we've run out of lengths to try, return a hash mismatch error.
+                // Otherwise, move on to the next longest object or next largest ignored padding.
+                self.mark_shortest_object_hash_invalid(data_hash, mapping, data)?;
+            }
+        }
+    }
+
+    /// Returns the maximum amount of data that still needs to be downloaded.
+    pub fn remaining_download_length(&self) -> usize {
```

Comment:
something like `largest_possible_object_length`?

Reply:
This calculation actually returns the largest amount of data that still needs to be downloaded (not the full length of the object). So I went with `max_remaining_download_length()`.

`maximum` seemed to be a shorter way of saying `largest_possible`.

Happy to change to `max_remaining_object_length()`, but that might be more confusing, because it doesn't say what it is "remaining" after.